# 5. Keywords Analysis

## 5.1 Task Setting and Description

**Objective**: Extract and analyze keywords from the list of publications using Python and public APIs.

**Tools**: 
- Python
- DOI lookup APIs (e.g., CrossRef API)
- Natural Language Processing libraries (e.g., NLTK, spaCy)
- Pandas for data handling
- Visualization libraries (e.g., matplotlib, seaborn, wordcloud, networkx)

---

## 5.2 Plan for Code Implementation

### 1. Data Retrieval

**Import Necessary Libraries**



In [3]:
!pip install pandas

  Using cached pandas-2.2.3-cp311-cp311-macosx_11_0_arm64.whl.metadata (89 kB)
  Using cached pytz-2024.2-py2.py3-none-any.whl.metadata (22 kB)
  Using cached tzdata-2024.2-py2.py3-none-any.whl.metadata (1.4 kB)
Using cached pandas-2.2.3-cp311-cp311-macosx_11_0_arm64.whl (11.3 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.4/5.4 MB 28.6 MB/s eta 0:00:00a 0:00:01
Using cached pytz-2024.2-py2.py3-none-any.whl (508 kB)
Using cached tzdata-2024.2-py2.py3-none-any.whl (346 kB)


In [4]:
!pip install requests

  Using cached requests-2.32.3-py3-none-any.whl.metadata (4.6 kB)
  Using cached charset_normalizer-3.4.0-cp311-cp311-macosx_11_0_arm64.whl.metadata (34 kB)
  Using cached idna-3.10-py3-none-any.whl.metadata (10 kB)
  Using cached urllib3-2.2.3-py3-none-any.whl.metadata (6.5 kB)
  Using cached certifi-2024.8.30-py3-none-any.whl.metadata (2.2 kB)
Using cached requests-2.32.3-py3-none-any.whl (64 kB)
Using cached certifi-2024.8.30-py3-none-any.whl (167 kB)
Using cached charset_normalizer-3.4.0-cp311-cp311-macosx_11_0_arm64.whl (118 kB)
Using cached idna-3.10-py3-none-any.whl (70 kB)
Using cached urllib3-2.2.3-py3-none-any.whl (126 kB)


In [ ]:
import pandas as pd
import requests



**Load DOIs from CSV Files**



In [6]:
# Load DOIs from CSV files in data/references folder
import pandas as pd

# List of file paths
file_paths = [
    '/Users/max/Documents/Code/tsi-sota-ai/data/references/1.2.2.1 LR - The Specialist Shortage and its Impact.csv',
    '/Users/max/Documents/Code/tsi-sota-ai/data/references/1.2.2.2 LR - AI Applications in SCM Decision Support.csv',
    '/Users/max/Documents/Code/tsi-sota-ai/data/references/1.2.2.3 LR - Human-AI Collaboration in SCM.csv',
    '/Users/max/Documents/Code/tsi-sota-ai/data/references/1.2.2.4 LR - Challenges and Limitations of LLMs in SCM.csv',
    '/Users/max/Documents/Code/tsi-sota-ai/data/references/1.2.2.5 LR - Decision-Making Processes.csv',
    '/Users/max/Documents/Code/tsi-sota-ai/data/references/1.2.2.6 LR - Agents.csv'
]

# Initialize an empty list to store DataFrames
dfs = []

# Loop through file paths, read each CSV, and add a column with the source file name
for file_path in file_paths:
    df = pd.read_csv(file_path)
    df['source_file'] = file_path.split('/')[-1]  # Extract the file name from the path
    dfs.append(df)

# Concatenate all DataFrames into a single DataFrame
references_df = pd.concat(dfs, ignore_index=True)

# Display the first few rows of the combined DataFrame
references_df.head()

date                                              title  \
0  2023-09-13  Transformative Procurement Trends: Integrating...   
1  2021-10-07  Exploring Progress with Supply Chain Risk Mana...   
2  2023-12-01  Exploring Applications and Practical Examples ...   
3  2021-09-27  Sustainable Innovations in the Food Industry t...   
4  2021-04-01  Artificial Intelligence (AI): Multidisciplinar...   

                               doi  \
0         10.3390/logistics7030063   
1         10.3390/logistics5040070   
2         10.3390/logistics7040091   
3         10.3390/logistics5040066   
4  10.1016/j.ijinfomgt.2019.08.002   

                                             authors  \
0  [{'author_name': 'Areej Althabatah', 'author_s...   
1  [{'author_name': 'Remko van Hoek', 'author_slu...   
2  [{'author_name': 'João Reis', 'author_slug': '...   
3  [{'author_name': 'Saurabh Sharma', 'author_slu...   
4  [{'author_name': 'Yogesh K. Dwivedi', 'author_...   

                                           journal  \
0                                        Logistics   
1                                        Logistics   
2                                        Logistics   
3                                        Logistics   
4  International Journal of Information Management   

                                     short_journal  volume  year    publisher  \
0                                        Logistics     7.0  2023      MDPI AG   
1                                        Logistics     5.0  2021      MDPI AG   
2                                        Logistics     7.0  2023      MDPI AG   
3                                        Logistics     5.0  2021      MDPI AG   
4  International Journal of Information Management    57.0  2021  Elsevier BV   

   issue    page                                           abstract  \
0    3.0      63  Background: the advent of Industry 4.0 (I4.0) ...   
1    4.0      70  Background: In response to calls for actionabl...   
2    4.0      91  Background: Material Requirements Planning (MR...   
3    4.0      66  The agri-food sector is an endless source of e...   
4    NaN  101994  As far back as the industrial revolution, sign...   

                                         source_file  
0  1.2.2.1 LR - The Specialist Shortage and its I...  
1  1.2.2.1 LR - The Specialist Shortage and its I...  
2  1.2.2.1 LR - The Specialist Shortage and its I...  
3  1.2.2.1 LR - The Specialist Shortage and its I...  
4  1.2.2.1 LR - The Specialist Shortage and its I...

In [7]:
references_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 593 entries, 0 to 592
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   date           593 non-null    object 
 1   title          593 non-null    object 
 2   doi            593 non-null    object 
 3   authors        593 non-null    object 
 4   journal        593 non-null    object 
 5   short_journal  563 non-null    object 
 6   volume         589 non-null    float64
 7   year           593 non-null    int64  
 8   publisher      593 non-null    object 
 9   issue          403 non-null    float64
 10  page           415 non-null    object 
 11  abstract       587 non-null    object 
 12  source_file    593 non-null    object 
dtypes: float64(2), int64(1), object(10)
memory usage: 60.4+ KB


In [8]:
# Export the combined DataFrame to a JSON file
references_df.to_json('/Users/max/Documents/Code/tsi-sota-ai/data/references_raw.json', orient='records', lines=True)



**Fetch Publication Metadata Using CrossRef API**



In [9]:
!pip install crossrefapi

  Attempting uninstall: urllib3
    Found existing installation: urllib3 2.2.3
    Uninstalling urllib3-2.2.3:
      Successfully uninstalled urllib3-2.2.3


In [10]:
from crossref.restful import Works

# Initialize the Works client
works = Works()

In [11]:
# Define a function to fetch metadata for a given DOI
def fetch_metadata(doi):
    try:
        # Fetch metadata for the given DOI
        metadata = works.doi(doi)
        return {
            'DOI': metadata.get('DOI'),
            'title': metadata.get('title', [None])[0],
            'authors': [author.get('given', '') + ' ' + author.get('family', '') for author in metadata.get('author', [])],
            'abstract': metadata.get('abstract'),
            'keywords': metadata.get('subject', [])
        }
    except Exception as e:
        print(f"Error fetching metadata for DOI {doi}: {e}")
        return None

In [12]:
# Example usage
doi = '10.1590/0102-311x00133115'
metadata = fetch_metadata(doi)
print(metadata)

{'DOI': '10.1590/0102-311x00133115', 'title': 'Congenital Zika virus syndrome', 'authors': ['Viroj Wiwanitki'], 'abstract': None, 'keywords': []}


In [17]:
# Loop through DOIs and collect metadata
!pip install tqdm
from tqdm import tqdm
import time

In [18]:
# Initialize the list to store metadata
metadata_list = []

# Iterate through each DOI in the references DataFrame with a 1-second delay
for doi in tqdm(references_df['doi'], desc='Fetching metadata'):
    metadata = fetch_metadata(doi)
    if metadata:
        metadata_list.append(metadata)
    time.sleep(1)  # Add a 1-second delay between requests

Fetching metadata: 100%|██████████| 593/593 [14:36<00:00,  1.48s/it]


In [19]:
# Create a DataFrame from the metadata list
references_metadata_crossref_df = pd.DataFrame(metadata_list)

# Display the first few rows of the metadata DataFrame
references_metadata_crossref_df.head()

DOI  \
0         10.3390/logistics7030063   
1         10.3390/logistics5040070   
2         10.3390/logistics7040091   
3         10.3390/logistics5040066   
4  10.1016/j.ijinfomgt.2019.08.002   

                                               title  \
0  Transformative Procurement Trends: Integrating...   
1  Exploring Progress with Supply Chain Risk Mana...   
2  Exploring Applications and Practical Examples ...   
3  Sustainable Innovations in the Food Industry t...   
4  Artificial Intelligence (AI): Multidisciplinar...   

                                             authors  \
0  [Areej Althabatah, Mohammed Yaqot, Brenno Mene...   
1                                   [Remko van Hoek]   
2                                        [João Reis]   
3  [Saurabh Sharma, Vijay Kumar Gahlawat, Kumar R...   
4  [Yogesh K. Dwivedi, Laurie Hughes, Elvira Isma...   

                                            abstract keywords  
0  <jats:p>Background: the advent of Industry 4.0...       []  
1  <jats:p>Background: In response to calls for a...       []  
2  <jats:p>Background: Material Requirements Plan...       []  
3  <jats:p>The agri-food sector is an endless sou...       []  
4                                               None       []

In [20]:
references_metadata_crossref_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 593 entries, 0 to 592
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   DOI       593 non-null    object
 1   title     593 non-null    object
 2   authors   593 non-null    object
 3   abstract  538 non-null    object
 4   keywords  593 non-null    object
dtypes: object(5)
memory usage: 23.3+ KB


As we can see, Crossref requests returned us empty lists for keywords, so let's try alternative approaches


**Fetch Publication Metadata Using Semantic Scholar API**

In [25]:
import requests
import time

# Function to fetch metadata using Semantic Scholar API with enhanced error handling and rate limiting
def fetch_semantic_scholar_metadata(doi, max_retries=3, backoff_factor=2):
    """
    Fetch metadata for a given DOI using the Semantic Scholar API.

    Args:
        doi (str): The DOI of the publication.
        max_retries (int): Maximum number of retries for failed requests.
        backoff_factor (int): Factor by which the wait time increases after each retry.

    Returns:
        dict or None: A dictionary containing metadata if successful, else None.
    """
    url = f'https://api.semanticscholar.org/graph/v1/paper/DOI:{doi}'
    params = {
        'fields': 'title,authors,abstract,fieldsOfStudy'
    }
    headers = {
        'User-Agent': 'YourProjectName/1.0 (your.email@example.com)'
    }
    
    for attempt in range(1, max_retries + 1):
        try:
            response = requests.get(url, params=params, headers=headers)
            
            if response.status_code == 200:
                data = response.json()
                return {
                    'DOI': doi,
                    'title': data.get('title'),
                    'authors': [author['name'] for author in data.get('authors', [])],
                    'abstract': data.get('abstract'),
                    'keywords': data.get('fieldsOfStudy', [])
                }
            elif response.status_code == 429:
                # Rate limit exceeded
                wait_time = backoff_factor ** attempt
                print(f"Rate limit exceeded. Retrying in {wait_time} seconds...")
                time.sleep(wait_time)
            else:
                print(f"Failed to fetch data for DOI {doi}. Status code: {response.status_code}")
                return None
        except requests.exceptions.RequestException as e:
            wait_time = backoff_factor ** attempt
            print(f"Error fetching data for DOI {doi}: {e}. Retrying in {wait_time} seconds...")
            time.sleep(wait_time)
    
    print(f"Failed to fetch data for DOI {doi} after {max_retries} attempts.")
    return None



In [23]:
# Example usage
doi = '10.1038/s41586-020-2649-2'
metadata = fetch_semantic_scholar_metadata(doi)
print(metadata)

{'DOI': '10.1038/s41586-020-2649-2', 'title': 'Array programming with NumPy', 'authors': ['Charles R. Harris', 'K. Millman', 'S. Walt', 'R. Gommers', 'Pauli Virtanen', 'D. Cournapeau', 'Eric Wieser', 'Julian Taylor', 'Sebastian Berg', 'Nathaniel J. Smith', 'Robert Kern', 'Matti Picus', 'Stephan Hoyer', 'M. Kerkwijk', 'M. Brett', 'A. Haldane', "Jaime Fern'andez del R'io", 'Marcy Wiebe', 'Pearu Peterson', "Pierre G'erard-Marchant", 'Kevin Sheppard', 'Tyler Reddy', 'Warren Weckesser', 'Hameer Abbasi', 'C. Gohlke', 'T. Oliphant'], 'abstract': None, 'keywords': ['Computer Science', 'Mathematics', 'Medicine']}


In [24]:
import pandas as pd
from tqdm import tqdm
import time

# Initialize the list to store metadata
metadata_list = []

# Iterate through each DOI in the references DataFrame with a 1-second delay
for doi in tqdm(references_df['doi'], desc='Fetching metadata'):
    metadata = fetch_semantic_scholar_metadata(doi)
    if metadata:
        metadata_list.append(metadata)
    time.sleep(1)  # Add a 1-second delay between requests to respect API rate limits

# Create a DataFrame from the metadata list
references_metadata_semanticscholar_df = pd.DataFrame(metadata_list)

# Export the metadata DataFrame to a JSON file
references_metadata_semanticscholar_df.to_json(
    '/Users/max/Documents/Code/tsi-sota-ai/data/references_metadata_semanticscholar.json',
    orient='records',
    lines=True
)

# Display the first few rows of the metadata DataFrame
references_metadata_semanticscholar_df.head()

Fetching metadata:   4%|▍         | 23/593 [00:35<14:59,  1.58s/it]

Failed to fetch data for DOI 10.1287/mnsc.2019.3499. Status code: 404


Fetching metadata:   9%|▊         | 51/593 [01:18<13:08,  1.45s/it]

Failed to fetch data for DOI 10.1287/mnsc.2019.3511. Status code: 404


Fetching metadata:  13%|█▎        | 75/593 [01:55<12:51,  1.49s/it]

Failed to fetch data for DOI 10.1287/mnsc.2023.4739. Status code: 404


Fetching metadata:  23%|██▎       | 134/593 [03:25<11:53,  1.56s/it]

Failed to fetch data for DOI 10.3390/smartcities6060142. Status code: 404


Fetching metadata:  25%|██▍       | 147/593 [03:45<11:21,  1.53s/it]

Failed to fetch data for DOI 10.3390/make5010017. Status code: 404


Fetching metadata:  25%|██▌       | 151/593 [03:51<11:23,  1.55s/it]

Failed to fetch data for DOI 10.3390/make5010017. Status code: 404


Fetching metadata:  30%|██▉       | 175/593 [04:30<10:46,  1.55s/it]


KeyboardInterrupt: 

We're getting 404 error, so let's wait for an API Key provisioning by Semantic Scholar. Meanwhile let's use third approach


**Extract Keywords from Abstracts**

Description: Use keyword extraction algorithms on available abstracts.

In [28]:
!pip install nltk

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 15.2 MB/s eta 0:00:00


In [29]:
import nltk
import string
from collections import Counter
from nltk.corpus import stopwords

In [30]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /Users/max/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [31]:
# Function to extract keywords from text
def extract_keywords(text, num_keywords=10):
    tokens = nltk.word_tokenize(text.lower())
    tokens = [word for word in tokens if word.isalpha()]
    filtered_tokens = [word for word in tokens if word not in stopwords.words('english')]
    word_counts = Counter(filtered_tokens)
    return [word for word, count in word_counts.most_common(num_keywords)]



---

### 2. Data Cleaning

**Import NLP Libraries**



In [ ]:
import nltk
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer

# Download NLTK data files if necessary
# nltk.download('stopwords')
# nltk.download('wordnet')



**Normalize Text**



In [ ]:
# Convert text to lowercase
# Remove stopwords
# Perform stemming or lemmatization
def clean_text(text):
    # Tokenize text
    # Remove stopwords
    # Lemmatize words
    return cleaned_text



---

### 3. Keyword Extraction

**Using TF-IDF or RAKE Algorithm**



In [ ]:
# Option 1: TF-IDF
from sklearn.feature_extraction.text import TfidfVectorizer

# Create a TF-IDF Vectorizer
tfidf_vectorizer = TfidfVectorizer(max_df=0.8, max_features=10000)

# Fit and transform the cleaned text data
tfidf_matrix = tfidf_vectorizer.fit_transform(cleaned_texts)

# Extract top keywords based on TF-IDF scores

# Option 2: RAKE Algorithm
# from rake_nltk import Rake

# Initialize RAKE
# rake_nltk_var = Rake()

# Extract keywords for each document



---

### 4. Visualization

**Generate Word Clouds or Bar Charts**



In [ ]:
import matplotlib.pyplot as plt
from wordcloud import WordCloud

# Generate a word cloud of keywords
wordcloud = WordCloud(background_color='white').generate_from_frequencies(keyword_frequencies)

# Display the generated image
plt.imshow(wordcloud, interpolation='bilinear')
plt.axis('off')
plt.show()



---

## 5.3 Infographics Generation

**Create Semantic Graphs of Keyword Relationships**



In [ ]:
import networkx as nx
import matplotlib.pyplot as plt

# Build a graph where nodes are keywords and edges represent co-occurrence
G = nx.Graph()

# Add nodes and edges based on keyword relationships

# Draw the network graph
pos = nx.spring_layout(G)
nx.draw(G, pos, with_labels=True, node_size=50, font_size=10)
plt.show()



**Use Visualization Libraries**

- Utilize `networkx` for building and visualizing network graphs.
- Use `matplotlib` or `seaborn` for enhanced visualization aesthetics.

---

**Note**: Replace placeholder code with actual implementations. Ensure all necessary data files are correctly referenced, and dependencies are installed in your environment.